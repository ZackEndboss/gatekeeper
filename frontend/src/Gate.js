import React from 'react';
import PropTypes from 'prop-types';
import R from 'ramda';
import {connect} from 'react-redux';
import {
    Button,
    Popover,
    OverlayTrigger,
    Grid,
    Row,
    Col,
    Glyphicon,
    Well
} from 'react-bootstrap';
import {setManualState, setCommentEditDialog} from './reducers/gates';
import {deleteTicketRequest} from './enhancers/api';
import CommentDialog from './CommentDialog';
import './Gate.css';

const manualStatePopover = (
    <Popover id="manualStatePopover" title='Manuel State'>
        "Close or open the gate manually. This state is persistent."
    </Popover>
);

export function ManuelState({isOpen, onManualStateClick}) {
    return (
        <OverlayTrigger placement="bottom" overlay={manualStatePopover}>
            <Button bsStyle={isOpen ? 'success' : 'danger'} onClick={() => onManualStateClick(!isOpen)}>
                {isOpen ? 'Open' : 'Closed'}
            </Button>
        </OverlayTrigger>
    );
}

export function LastModified({isOpen, last_modified}) {
    return isOpen ? null : <div className='default-comment'>{last_modified}</div>;
}

const autoStatePopover = (
    <Popover id="autoStatePopover" title='Automatic gate'>
        "Will be temporally closed if tickets are present, on holidays or outside of business hours."
    </Popover>
);

export function AutoState({isOpen}) {
    return (
        <OverlayTrigger placement="bottom" overlay={autoStatePopover}>
            <div className='auto-state-button-div'>
                <Button className='auto-state-button' disabled bsStyle={isOpen ? 'success' : 'danger'}>
                    {isOpen ? 'Open' : 'Closed'}
                </Button>
            </div>
        </OverlayTrigger>
    );
}

export function GateArrow({gate_state}) {
    return <Glyphicon className={gate_state ? 'text-danger' : 'text-success'} glyph='chevron-right'/>;
}

export function GateButton() {

}

export function Comment({comment, openCommentEditDialog}) {
    return (
        <div className={`comment ${comment ? '' : 'default-comment'}`}>
            {comment ? comment : 'comment'}
            <a className='glyp-button'
               onClick={() => openCommentEditDialog()}>
                <Glyphicon glyph='pencil'/>
            </a>
        </div>
    );
}

export function Tickets({tickets, onTicketRemoveClick}) {
    return (
        <Grid>
            {tickets.map(ticket => {
                return (
                    <Row key={ticket}>
                        <Col>
                            {ticket}
                            <a className='glyp-button'
                               onClick={() => onTicketRemoveClick(ticket)}>
                                <Glyphicon glyph='trash'/>
                            </a>
                        </Col>
                    </Row>
                );
            })}
        </Grid>
    );
}

export class Gate extends React.Component {
    render() {
        const {manual_state, auto_state, message, queue, last_modified, onTicketRemoveClick, onManualStateClick, openCommentEditDialog} = this.props;
        const gate_state = !manual_state & auto_state;
        return (
            <Well className="gate" bsSize="small">
                <div>
                    <GateArrow gate_state={gate_state}/>
                    <ManuelState isOpen={manual_state} onManualStateClick={onManualStateClick}/>
                    <GateArrow gate_state={gate_state}/>
                    <AutoState isOpen={auto_state}/>
                    <GateArrow gate_state={gate_state}/>
                </div>
                <LastModified isOpen={manual_state} last_modified={last_modified}/>
                <div>
                    <CommentDialog group={this.props.group}
                                   service={this.props.service}
                                   environment={this.props.environment}
                                   comment={message}/>

                    <Comment comment={message} openCommentEditDialog={openCommentEditDialog}/>
                </div>
                <Tickets tickets={queue} onTicketRemoveClick={onTicketRemoveClick}/>
            </Well>
        );
    }
}

Gate.propTypes = {
    group: PropTypes.string.isRequired,
    service: PropTypes.string.isRequired,
    environment: PropTypes.string.isRequired,
    message: PropTypes.string,
    manual_state: PropTypes.bool,
    auto_state: PropTypes.bool,
    queue: PropTypes.array,
    last_modified: PropTypes.string
};

const defaultValues = {
    group: '',
    service: '',
    environment: '',
    message: '',
    manual_state: true,
    auto_state: true,
    queue: [],
    last_modified: 'now'
};

const mapStateToProps = (state, initialProps) => {
    const gate_from_state = R.path(['gates', initialProps.group, initialProps.service, initialProps.environment])(state);
    return R.mergeAll([defaultValues, initialProps, gate_from_state]);
};

const mapDispatchToProps = (dispatch, initialProps) => {
    return {
        onTicketRemoveClick: (id) => {
            dispatch(deleteTicketRequest(initialProps.group, initialProps.service, initialProps.environment, id));
        },
        onManualStateClick: (state) => {
            dispatch(setManualState(initialProps.group, initialProps.service, initialProps.environment, state));
        },
        openCommentEditDialog: () => {
            dispatch(setCommentEditDialog(initialProps.group, initialProps.service, initialProps.environment, true));
        }
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(Gate);
