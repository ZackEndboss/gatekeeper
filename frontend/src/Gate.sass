.gate
  margin: 10px

.comment
  text-align: right
  white-space: pre-wrap
  white-space: -moz-pre-wrap
  //white-space: -pre-wrap
  white-space: -o-pre-wrap
  word-wrap: break-word

.default-comment
  font-style: italic
  color: gray

.auto-state-button-div
  display: inline-block
  cursor: not-allowed

.auto-state-button
  pointer-events: none