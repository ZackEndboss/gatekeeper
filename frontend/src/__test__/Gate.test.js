import React from 'react';
import {Gate, ManuelState, AutoState, Comment, Tickets}  from '../Gate';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';

describe('Gate component', () => {
    it('should contain all components', () => {
        const component = renderer.create(<Gate group='group'
                                                service='service'
                                                environment='environment'
                                                comment='comment 1'
                                                manual_state={true}
                                                auto_state={true}
                                                tickets={['ticket 1', 'ticket 2']}
                                                last_modified_state='some time ago'
                                                last_modified_comment='a while ago'/>);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('the manual gate state should be closed', () => {
        const component = mount(<ManuelState isOpen={false}/>);
        expect(component.text()).toEqual('Closed');
        expect(component.find('Button').props().bsStyle).toEqual('danger');
    });

    it('the manual gate state should be open', () => {
        const component = mount(<ManuelState isOpen={true}/>);
        expect(component.text()).toEqual('Open');
        expect(component.find('Button').props().bsStyle).toEqual('success');
    });

    it('the auto gate state should be closed', () => {
        const component = mount(<AutoState isOpen={false}/>);
        expect(component.text()).toEqual('Closed');
        expect(component.find('Label').props().bsStyle).toEqual('danger');
    });

    it('the auto gate state should be open', () => {
        const component = mount(<AutoState isOpen={true}/>);
        expect(component.text()).toEqual('Open');
        expect(component.find('Label').props().bsStyle).toEqual('success');
    });

    it('the comment field contains the given comment', () => {
        const component = mount(<Comment comment='some comment'/>);
        expect(component.text()).toEqual('some comment');
    });

    it('all tickets are displayed', () => {
        const component = mount(<Tickets tickets={['ticket 1', 'ticket 2']}/>);
        const tickets = component.find('Row');
        expect(tickets.length).toEqual(2);
        expect(tickets.at(0).text()).toEqual('ticket 1');
        expect(tickets.at(1).text()).toEqual('ticket 2');
    });

});