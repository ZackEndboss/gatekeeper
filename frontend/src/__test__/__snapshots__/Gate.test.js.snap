// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gate component should contain all components 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-md-1 col-xs-1"
    >
      <div
        className="row"
      >
        <button
          className="btn btn-success"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Open
        </button>
      </div>
      <div
        className="row"
      >
        <div>
          some time ago
        </div>
      </div>
      <div
        className="row"
      >
        <h4>
          <span
            className="label label-success"
          >
            Open
          </span>
        </h4>
      </div>
    </div>
    <div
      className="col-md-2 col-xs-2"
      style={
        Object {
          "marginLeft": "-45px",
        }
      }
    >
      <div
        className="form-group"
      >
        <textarea
          className="form-control"
          id="formControlsTextarea"
          onChange={[Function]}
          onClick="openCommentEditDialog"
          type={undefined}
          value="comment 1"
        />
      </div>
      <button
        className="btn btn-xs btn-info"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        edit comment
      </button>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className=""
    >
      <div
        className="container"
      >
        <div
          className="row"
        >
          <div
            className=""
          >
            ticket 1
            <a
              onClick={[Function]}
            >
              <span
                className="glyphicon glyphicon-trash"
              />
            </a>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className=""
          >
            ticket 2
            <a
              onClick={[Function]}
            >
              <span
                className="glyphicon glyphicon-trash"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`load from state should load the state correctly 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-md-1 col-xs-1"
    >
      <div
        className="row"
      >
        <button
          className="btn btn-danger"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Closed
        </button>
      </div>
      <div
        className="row"
      >
        <div>
          now
        </div>
      </div>
      <div
        className="row"
      >
        <h4>
          <span
            className="label label-success"
          >
            Open
          </span>
        </h4>
      </div>
    </div>
    <div
      className="col-md-2 col-xs-2"
      style={
        Object {
          "marginLeft": "-45px",
        }
      }
    >
      <div
        className="form-group"
      >
        <textarea
          className="form-control"
          id="formControlsTextarea"
          onChange={[Function]}
          onClick={[Function]}
          type={undefined}
          value="some comment"
        />
      </div>
      <button
        className="btn btn-xs btn-info"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        edit comment
      </button>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className=""
    >
      <div
        className="container"
      >
        <div
          className="row"
        >
          <div
            className=""
          >
            ticket 1
            <a
              onClick={[Function]}
            >
              <span
                className="glyphicon glyphicon-trash"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
